### DISCUSSION TITLE
net-core-multi-stage-dockerfile-with-test-and-code-coverage-in-azure-pipelines/

### DISCUSSION THREAD

**Discussion by giscus** [colindembovsky/colindembovsky.github.io#19] [Giscussions]
# net-core-multi-stage-dockerfile-with-test-and-code-coverage-in-azure-pipelines/

All things DevOps and GitHub. Musings about DevOps tooling, culture and philosophy.

https://colinsalmcorner.com/net-core-multi-stage-dockerfile-with-test-and-code-coverage-in-azure-pipelines/

**Comment by divekarvinit** [comment]
Hello Colin, thanks for the detailed blog. I had a issue that after the report was published, when I went to individual files, it showed error as: File '/app/x_y_z.cs' does not exist (any more). I found out that it was due to WORKDIR in docker was set as `/app`.  With that, the source code is copied into `/app` directory. Now when report generator is run, the source has the path starting with `/app`.

Now, in azure pipeline, when the line `docker cp testcontainer:/testresults ./testresults` is run, it copies the generated report to the following path, `/home/vsts/work/1/s/TestResults`. The report tries to find files starting at `/app` but cannot find. Hence we see the error.

To get around I had to update my Dockerfile's build stage's WORKDIR as `/home/vsts/work/1/s` and update your script to copy source code in azure host as:
```
- script: |
    export id=$(docker images --filter "label=test=true" -q | head -1)
    docker create --name testcontainer $id
    docker cp testcontainer:/testresults ./testresults
    docker cp testcontainer:$(System.DefaultWorkingDirectory) $(System.DefaultWorkingDirectory)
    docker rm testcontainer
  displayName: 'get test results'

```

My Docker file looks like:
```
FROM mcr.microsoft.com/dotnet/sdk:6.0 AS build
WORKDIR /home/vsts/work/1/s
COPY . .

# Ommitted for convinience.
```

With this change, at both places (in docker and azure host), the source code lies at `/home/vsts/work/1/s`.

I feel that this work-around solution could be improved. Would you have any idea?

**Comment by Gladskih** [comment]
But if a test failed the image (and layer) is not built and the report is lost.
